# 01. Exercise from course

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import data sets
path = r'/Users/minglunchen/Desktop/12-2023 Instacart Basket Analysis'

In [3]:
path

'/Users/minglunchen/Desktop/12-2023 Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [46]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  orders_interval  
count       3.421083e+06     3.214874e+06  
mean        1.345202e+01     1.111484e+01  
std         4.226088e+00     9.206737e+00  
min         0.000000e+00     0.000000e+00  
25%         1.000000e+01     4.000000e+00  
50%         1.300000e+01     7.000000e+00  
75%         1.600000e+01     1.500000e+01  
max         2.300000e+01     3.000000e+01

In [8]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
# 1. Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
## Check for mixed type
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [13]:
### Fix mixed-type columns
df_test['mix'] = df_test['mix'].astype('str')

In [5]:
# 2. Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
## Create a subset of the dataframe containing those values in question
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [7]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [8]:
### Addressing Missing Values -- 3 methods mentioned in previous courses - Method 2: Impute value with mean or median of the column --- choose mean using describe(), choose median using df_prods.median().

In [9]:
### As an example using the following code when column with missing is numeric: df_prods['column with missing'].fillna(mean value, inplace=True) or df_prods['column with missing'].fillna(median value, inplace=True)

In [10]:
### Since missing value are string, we can either remove the missing values or create a subset dataframe for those not missing ones.
df_prods.shape

(49693, 5)

In [11]:
#### Create a new datafrom for clean values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#### Or we can drop all missing values
df_prods.dropna(inplace = True)
#### Drop only NaNs from a particular column
df_prods.dropna(subset = ['product_name'], inplace = True)

In [13]:
# 3. Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
## Addressing Duplicates
df_prods_clean.shape

(49677, 5)

In [16]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
df_prods_clean_no_dups.shape

(49672, 5)

In [18]:
# Export clean data -- products
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 02. Task

In [43]:
# Step 2: Run the df.describe() to interpret the output.
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  orders_interval  
count       3.421083e+06     3.214874e+06  
mean        1.345202e+01     1.111484e+01  
std         4.226088e+00     9.206737e+00  
min         0.000000e+00     0.000000e+00  
25%         1.000000e+01     4.000000e+00  
50%         1.300000e+01     7.000000e+00  
75%         1.600000e+01     1.500000e+01  
max         2.300000e+01     3.000000e+01

Answers for Step 2: 
The count of the column 'orders_interval' is less than the other columns, indicating missing values in this column.

In [47]:
# Step 3: Check for mixed-type data in df_ords dataframe.
# Step 4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

Answers for Step 3&4:
There is no mixed-type data in df_ords dataframe.

In [48]:
# Step 5: Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()

order_id                   0
user_id                    0
eval_set                   0
order_number               0
orders_day_of_week         0
order_hour_of_day          0
orders_interval       206209
dtype: int64

Answers for Step 5:
There are 206209 missing values in the column 'orders_interval' (which is 'days_since_prior_order'). 
The potential explanation: these 206209 orders might be the first order, that means there was no prior order ahead.

In [49]:
# To verify the hypothese, a subset of the dataframe containing those values in question was created.
df_ords_interval_null = df_ords[df_ords['orders_interval'].isnull() == True]

In [50]:
df_ords_interval_null

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
11        2168274        2    prior             1                   2   
26        1374495        3    prior             1                   1   
39        3343014        4    prior             1                   6   
45        2717275        5    prior             1                   3   
...           ...      ...      ...           ...                 ...   
3420930    969311   206205    prior             1                   4   
3420934   3189322   206206    prior             1                   3   
3421002   2166133   206207    prior             1                   6   
3421019   2227043   206208    prior             1                   1   
3421069   3154581   206209    prior             1                   3   

         order_hour_of_day  orders_interval  
0                        8              NaN  
11                      11              NaN  
26                      14              NaN  
39                      11              NaN  
45                      12              NaN  
...                    ...              ...  
3420930                 12              NaN  
3420934                 18              NaN  
3421002                 19              NaN  
3421019                 15              NaN  
3421069                 11              NaN  

[206209 rows x 7 columns]

Answers for Step 5 cont.:
After identifying the missing values in the column 'orders_interval', I found the 'order_number' of these items were all 1, indicating they were the first order -- and my hypothese above was correct.

In [51]:
# Step 6: Address the missing values using an appropriate method.
df_ords['first_order'] = df_ords['orders_interval'].isnull() == True

In [52]:
df_ords

order_id  user_id eval_set  order_number  orders_day_of_week  \
0         2539329        1    prior             1                   2   
1         2398795        1    prior             2                   3   
2          473747        1    prior             3                   3   
3         2254736        1    prior             4                   4   
4          431534        1    prior             5                   4   
...           ...      ...      ...           ...                 ...   
3421078   2266710   206209    prior            10                   5   
3421079   1854736   206209    prior            11                   4   
3421080    626363   206209    prior            12                   1   
3421081   2977660   206209    prior            13                   1   
3421082    272231   206209    train            14                   6   

         order_hour_of_day  orders_interval  first_order  
0                        8              NaN         True  
1                        7             15.0        False  
2                       12             21.0        False  
3                        7             29.0        False  
4                       15             28.0        False  
...                    ...              ...          ...  
3421078                 18             29.0        False  
3421079                 10             30.0        False  
3421080                 12             18.0        False  
3421081                 12              7.0        False  
3421082                 14             30.0        False  

[3421083 rows x 8 columns]

Answers for Step 6:
For such a large volumn of missing values, it was not wise to simply remove them. I therefore chose to address the missing values by creating a new variable that acts like a flag. The new variable was named 'first_order', and 'True' indicated this was the first order for the user.

In [53]:
# Step 7: Run a check for duplicate values in your df_ords data.
# Step 8: Address the duplicates using an appropriate method.
df_ords_dups = df_ords[df_ords.duplicated()]

In [54]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, orders_interval, first_order]
Index: []

Answers for Step 7&8:
No duplicate values were found in the df_ords data.

In [55]:
# Step 9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))